citest/gcp_testing/resource_snapshot.py

```diff
@@ -623,13 +651,78 @@ def transform(items):
 
+  def __list_resource_helper(self, agent, method_name, resource, transform):
+    """Helper function for listing resources.
+
+    This function will associate the parameters used to invoke the function
+    with the results from the function so that the listed results can be
+    operated on again in the future (e.g. delete) where additional parameters
+    for their context might be needed (e.g. project, zone, etc).
+
+    Normally this will either call list or listAggregated
+    (for some Compute APIs). The listAggregated response is actually a
+    composite response containing a dictionary keyed by the parameter
+    aggregated over (e.g. zone) and whose value is the result list for that
+    particular zone. This method unpacks that dictionary into a list of
+    tuples where the first element is the key (e.g. identifying the zone)
+    and the second element is the result value (the inner item data list.
+    For normal 'list' methods, this returns just the non-tupled item list.
+    The tuple form communicates an additional [implied] parameter that will
+    be needed to retrieve or operate on the item.
+
+    This module introduces some fake "method_name" parameters for synthetic
+    aggregation. If the method is LIST_BY_ZONE then this function will manually
+    iterate over each ZONE and perform a 'list' operation, but return the
+    results from all these queries as if it were a single 'listAggregated'
+    call. This is to support APIs that do not have a listAggregated method
+    but require a zone or region parameter.
+
+    The method that calls this figures out whether or not it should iterate
+    over zones or regions when passing i the "method name". If one of these
+    synthetic methods is passed in, the result will report as if it came
+    from 'aggregatedList' even though this method does not actually exist.
+    This is so other functions dont need additional special cases since the
+    result schema is the same as if there were an "aggregatedList".
+    """
+
+    context = ExecutionContext()
+
+    if method_name == LIST_BY_ZONE:
+      param_name = 'zone'
+      param_variants = get_all_zones()
+    elif method_name == LIST_BY_REGION:
+      param_name = 'region'
+      param_variants = get_all_regions()
+    else:
+      params = agent.resource_method_to_variables(method_name, resource)
+      return params, method_name, agent.list_resource(
+          context, resource,
+          method_variant=method_name,
+          item_list_transform=transform)
+
+    params = agent.resource_method_to_variables(
+        'list', resource, **{param_name: 'tbd'})
+    items = []
+    for param_value in param_variants:
+      results = agent.list_resource(
+          context, resource, method_variant='list',
+          **{param_name: param_value})
+      items.extend([('{0}s/{1}'.format(param_name, param_value), item)
+                    for item in results])
+
+    return params, 'aggregatedList', items
```

Comment:
I'd wrap this tuple in an object. It might seem superfluous, but unnamed tuples aren't maintainable. With a class, the code will document itself.